### PR TITLE
8319598: SMFParser misinterprets interrupted running status

### PR DESCRIPTION
**Pull request by truj**
The MIDI file parser misinterprets events without status byte when they appear directly after a Meta of SysEx event.

For my bugfix I had to decide between two possible solutions:
- Strict solution: Throw an InvalidMidiDataException
- Tolerant solution: Use the status of the last channel event as running status

I used the tolerant solution.
I will send an email to the list client-libs-dev with my reasons.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319598](https://bugs.openjdk.org/browse/JDK-8319598): SMFParser misinterprets interrupted running status (**Bug** - P3)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16546/head:pull/16546` \
`$ git checkout pull/16546`

Update a local copy of the PR: \
`$ git checkout pull/16546` \
`$ git pull https://git.openjdk.org/jdk.git pull/16546/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16546`

View PR using the GUI difftool: \
`$ git pr show -t 16546`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16546.diff">https://git.openjdk.org/jdk/pull/16546.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16546#issuecomment-1799773019)

### CODE REVIEW

**Comment by truj** [comment]
I should not have written my reasons as an Email but directly as a Comment in Github.
Now the statistics part is nearly unreadable.
So only this part again but correctly formatted:

**Summary:**

- 1.1% of all MIDI files fail to be loaded by SMFParser (for ANY reason)
- 0.43% of all MIDI files contain interrupted running status (maybe some more that have a different problem before the first IRS occurs)
- From the files with IRS
  - before the bugfix:
    - 15.67% can be parsed without exception (but with incorrectly interpreted data)
    - 69.9% fail with InvalidMidiDataException
    - 14.42% fail with EOFException
  - after the bugfix:
    - 94.2% can be parsed without exception
    - 5.4% fail with InvalidMidiDataException
    - 0.4% fail with EOFException
- The bugfix fixes exceptions for:
  - 78.5% of files with interrupted running status
  - 30.25% of ALL failing MIDI files in general !!!!!
- The bugfix doesn't change anything to any MIDI file without interrupted running status

**Details:**

Statistics from (only) the 721 files with interrupted running status:
- before the fix:
  - all: 721
  - success: 113
  - InvalidMidiDataException: 504
  - EOFException: 104
- after the fix:
  - all: 721
  - success: 679
  - InvalidMidiDataException: 39
  - EOFException: 3

Statistics from ALL files before and after the fix:
- before the fix:
  - all: 169457
  - success: 167586
  - fail: 1871
    - InvalidMidiDataException: 1606
    - EOFException: 265
- after the fix:
  - all: 169457
  - success: 168152
  - fail: 1305
    - InvalidMidiDataException: 1141
    - EOFException: 164

File: test/jdk/javax/sound/midi/File/SMFInterruptedRunningStatus.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        // check status byte of each message
+        int[] expectedStatusBytes = new int[] {0x90, 0xFF, 0x90, 0x90, 0x90, 0xFF, 0xFF};
+        for (int i = 0; i < expectedStatusBytes.length; i++) {
+            int expected = expectedStatusBytes[i];
+            if (expected != track.get(i).getMessage().getStatus()) {
+                throw new RuntimeException("file number " + fileNumber + " fails (wrong status byte in event " + i + ")");
+            }
+        }
+    }
+
```
Comment (mrserb) [inline test/jdk/javax/sound/midi/File/SMFInterruptedRunningStatus.java:70]: please split the long line above to have 80 chars per line.

The code below for "SMF_1"/etc are fine as is.

Comment (truj) [reply test/jdk/javax/sound/midi/File/SMFInterruptedRunningStatus.java:77]: Thank you for the suggestion. I reduced the line lengths.
Also for the code below SMF_..., even if you think it was fine.
I think now it's even more readable. But I can redo this part, if you think it was better before.

**Comment by mrserb** [comment]
keep open

**Review by prrace — APPROVED** [review]
This all sounds good to me but @mrserb should approve too since he was looking at it.

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**